src/Dictionaries/HashedDictionary.cpp

```diff
@@ -1023,6 +1024,8 @@ void HashedDictionary<dictionary_key_type, sparse, sharded>::calculateBytesAlloc
 
     for (const auto & arena : string_arenas)
         bytes_allocated += arena->allocatedBytes();
+
+    bucket_count = new_bucket_count;
```

Comment:
Why don't we reset `bytes_allocated` in the same way?

Reply:
Because `bytes_allocated` should be a sum across all attributes, while `bucket_count` should not (since it is used to calculate the `load_factor` as `element_count / bucket_count`)

But your question made me look into this deeply, and I found that `element_count` was incorrectly calculated for dictionaries with multiple attributes, so I've rebased the patch set, add some comments and update the test.